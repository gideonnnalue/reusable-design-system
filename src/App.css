* {
  margin: 0;
  padding: 0;
}

body {
  font-family: -apple-system, "Segoe UI", "Roboto", "Arial", sans-serif;
  color: #191d2f;
}

.checkbox-container {
  display: flex;
  justify-content: flex-end;
  padding: 20px;
}

.select-wrapper {
  display: flex;
}
