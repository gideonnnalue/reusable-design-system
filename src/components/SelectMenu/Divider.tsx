import { FC } from "react";
import { SelectMenuInput } from "../../theme/components";

const {
  SelectMenuItem: { MenuItemDivider },
} = SelectMenuInput;

const Divider: FC = (): JSX.Element => {
  return <MenuItemDivider />;
};

export default Divider;
