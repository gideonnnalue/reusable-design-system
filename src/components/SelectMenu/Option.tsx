import { useEffect } from "react";
import checkImg from "../../images/check-light.svg";

import { SelectMenuInput } from "../../theme/components";

import { OptionProps } from "./types";

const {
  SelectMenuItem: { MenuItemWrapper },
} = SelectMenuInput;

const Option = ({
  children,
  defaultValue,
  value,
  index,
  activeIndex,
  currentValue,
  onItemClicked,
}: OptionProps): JSX.Element => {
  useEffect(() => {
    if (defaultValue && defaultValue === value) {
      onItemClicked(value, index, children);
    }
  }, []);
  return (
    <MenuItemWrapper
      selected={currentValue === value && index === activeIndex}
      role="option"
      aria-selected={currentValue === value && index === activeIndex}
      aria-disabled={false}
      data-value={value}
      tabIndex={currentValue === value && index === activeIndex ? 0 : -1}
      onClick={() => onItemClicked(value, index, children)}
    >
      <span>
        <img src={checkImg} alt="check" />
      </span>
      {children}
    </MenuItemWrapper>
  );
};

export default Option;
