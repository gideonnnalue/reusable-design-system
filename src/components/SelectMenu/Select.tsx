import React, { Fragment, useEffect, useState, useRef, FC } from "react";

import chevronImg from "../../images/chevron-down.svg";

import { SelectMenuInput } from "../../theme/components";
import { SelectProps } from "./types";

const {
  SelectMenu: {
    SelectMenuWrapper,
    SelectMenuContainer,
    SelectMenuDropdownWrapper,
    SelectMenuDropdown,
  },
} = SelectMenuInput;

const Select: FC<SelectProps> = ({
  children,
  defaultValue,
  onChange,
}): JSX.Element => {
  const node: any = useRef();
  const [currentValue, setCurrentValue] = useState(defaultValue);
  const [activeText, setActiveText] = useState("");
  const [activeIndex, setActiveIndex] = useState(0);
  const [selectOpen, setSelectOpen] = useState(false);
  const onItemClicked = (value: string, index: number, text: string) => {
    setCurrentValue(value);
    setActiveIndex(index);
    setActiveText(text);
    onChange(value, index);
  };

  const childrenModified = React.Children.map(children, (child, index) => {
    return React.cloneElement(child, {
      onItemClicked,
      index,
      currentValue,
      activeIndex,
      defaultValue,
    });
  });

  const openSelect = () => {
    setSelectOpen(true);
  };
  const closeSelect = () => {
    setSelectOpen(false);
  };

  const handleClickOutside = (e: any) => {
    if (node.current.contains(e.target)) {
      return;
    }
    closeSelect();
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  return (
    <SelectMenuWrapper data-testid="select-wrapper" ref={node}>
      <SelectMenuContainer
        data-testid="select-button"
        role="button"
        tabIndex={0}
        onClick={openSelect}
      >
        <span data-testid="selected-text">
          {activeText || "Select an Item"}
        </span>
        <img src={chevronImg} alt="arrow" />
      </SelectMenuContainer>
      <SelectMenuDropdownWrapper open={selectOpen}>
        <SelectMenuDropdown role="listbox">
          <Fragment>{childrenModified}</Fragment>
        </SelectMenuDropdown>
      </SelectMenuDropdownWrapper>
    </SelectMenuWrapper>
  );
};

export default Select;
