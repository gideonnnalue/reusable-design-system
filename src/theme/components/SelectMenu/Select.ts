import styled from "@emotion/styled";
import css from "@styled-system/css";

import { MenuDropdownProps } from "./types";
import { ThemeInterface } from "../../types";

const SelectMenuWrapper = styled.div(
  css(({ space }: ThemeInterface) => {
    return {
      margin: space["s-25"],
      position: "relative",
      display: "inline-block",
    };
  })
);

const SelectMenuContainer = styled.div(
  css(({ borders, colors, borderRadius, space, fontSizes }: ThemeInterface) => {
    return {
      border: borders.thin,
      borderColor: colors.gray[300],
      borderRadius: borderRadius.small,
      display: "flex",
      minWidth: 100,
      width: "fit-content",
      overflow: "hidden",
      position: "relative",
      cursor: "pointer",
      "& span": {
        display: "block",
        paddingX: space["s-15"],
        paddingY: space["s-10"],
        fontSize: fontSizes[1],
      },
      "&::after": {
        position: "absolute",
        content: '""',
        bottom: 0,
        left: 0,
        width: "100%",
        height: 1,
        backgroundColor: colors.gray[200],
      },
      "& img": {
        paddingRight: space["s-10"],
      },
    };
  })
);

const SelectMenuDropdownWrapper = styled.div(({ open }: MenuDropdownProps) =>
  css(({ transition }: ThemeInterface) => ({
    position: "absolute",
    visibility: open ? "visible" : "hidden",
    whiteSpace: "nowrap",
    opacity: open ? 1 : 0,
    left: 0,
    top: 39,
    transform: open ? "scale(1)" : "scale(0)",
    transition: transition.dropdown,
    transformOrigin: "0px 0px",
    zIndex: 1,
  }))
);

const SelectMenuDropdown = styled.ul(
  css(({ colors, borders, borderRadius, boxShadow }: ThemeInterface) => ({
    backgroundColor: colors[100],
    listStyle: "none",
    margin: 0,
    padding: 0,
    border: borders.thin,
    borderColor: colors.gray[300],
    borderRadius: borderRadius.small,
    minWidth: 100,
    boxShadow: boxShadow,
  }))
);

export {
  SelectMenuWrapper,
  SelectMenuContainer,
  SelectMenuDropdownWrapper,
  SelectMenuDropdown,
};
