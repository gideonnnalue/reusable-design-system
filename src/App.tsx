import { FC } from "react";
import { ThemeProvider } from "@emotion/react";
import "./App.css";

import { SelectMenu } from "./components";
import theme from "./theme";

const { Select, Option, OptionExtended, Divider } = SelectMenu;

const App: FC = () => {
  return (
    <ThemeProvider theme={theme}>
      <div className="">
        <Select defaultValue="" onChange={() => {}}>
          <OptionExtended
            value="hopper"
            desc="Grace Hopper was an American computer scientist and US navy rear admiral"
          >
            Hopper
          </OptionExtended>
          <OptionExtended
            value="holderton"
            desc="Frances Elizabeth Holberton was one of the programmers of the first computer"
          >
            Holberton
          </OptionExtended>
          <Divider />
          <OptionExtended
            value="teitelbaum"
            desc="Ruth Teitelbaum Hopper was one of the first computer programmers of the world"
          >
            Teitelbaum
          </OptionExtended>
        </Select>

        {/* <div>
          <Select defaultValue="" onChange={() => {}}>
            <Option value="hopper">
              Hopperjdjjjdjdjdjdjjdjddjjddjdpspoopdsopdosdsklsldklskldlksldksldkjjdjjj
            </Option>
            <Option value="holberton">Holberton</Option>
            <Divider />
            <Option value="antonelli">Antonelli</Option>
            <Option value="bartik">Bartik</Option>
            <Option value="teitelbaum">Teitelbaum</Option>
            <Option value="bartik">Bartik</Option>
          </Select>
        </div> */}
      </div>
    </ThemeProvider>
  );
};

export default App;
