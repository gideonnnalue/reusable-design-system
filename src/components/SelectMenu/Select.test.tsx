import React from "react";
import { render, cleanup } from "@testing-library/react";

import Select from "./Select";
import Option from "./Option";
import Divider from "./Divider";
import App from "../../App";

// afterEach(cleanup);

it("should take a snapshot", () => {
  const { asFragment } = render(
    <App>
      <Select defaultValue="" onChange={() => {}}>
        <Option value="hopper">Hopper</Option>
        <Option value="holberton">Holberton</Option>
        <Divider />
        <Option value="antonelli">Antonelli</Option>
        <Option value="bartik">Bartik</Option>
        <Option value="teitelbaum">Teitelbaum</Option>
        <Option value="bartik">Bartik</Option>
      </Select>
    </App>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have a role attribute", () => {
  const { getByTestId } = render(
    <App>
      <Select defaultValue="" onChange={() => {}}>
        <Option value="hopper">Hopper</Option>
        <Option value="holberton">Holberton</Option>
        <Divider />
        <Option value="antonelli">Antonelli</Option>
        <Option value="bartik">Bartik</Option>
        <Option value="teitelbaum">Teitelbaum</Option>
        <Option value="bartik">Bartik</Option>
      </Select>
    </App>
  );

  expect(getByTestId("select-button")).toHaveAttribute("role");
});
