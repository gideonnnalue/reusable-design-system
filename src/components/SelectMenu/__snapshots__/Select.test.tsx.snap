// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="css-1hnyd0r"
      data-testid="select-wrapper"
    >
      <div
        class="css-fdvb0z"
        data-testid="select-button"
        role="button"
        tabindex="0"
      >
        <span>
          Select an Item
        </span>
        <img
          alt="arrow"
          src="chevron-down.svg"
        />
      </div>
      <div
        class="css-1vznqvx"
      >
        <ul
          class="css-5ubqti"
          role="listbox"
        >
          <li
            aria-disabled="false"
            aria-selected="false"
            class="css-k4di0t"
            data-value="hopper"
            role="option"
            tabindex="-1"
          >
            <img
              alt="check"
              src="check-light.svg"
            />
            <span>
              Hopper
              <p
                class="css-1bj4mrz"
              >
                Grace Hopper was an American computer scientist and US navy rear admiral
              </p>
            </span>
          </li>
          <li
            aria-disabled="false"
            aria-selected="false"
            class="css-k4di0t"
            data-value="holderton"
            role="option"
            tabindex="-1"
          >
            <img
              alt="check"
              src="check-light.svg"
            />
            <span>
              Holberton
              <p
                class="css-1bj4mrz"
              >
                Frances Elizabeth Holberton was one of the programmers of the first computer
              </p>
            </span>
          </li>
          <div
            class="css-1trqa8a"
          />
          <li
            aria-disabled="false"
            aria-selected="false"
            class="css-k4di0t"
            data-value="teitelbaum"
            role="option"
            tabindex="-1"
          >
            <img
              alt="check"
              src="check-light.svg"
            />
            <span>
              Teitelbaum
              <p
                class="css-1bj4mrz"
              >
                Ruth Teitelbaum Hopper was one of the first computer programmers of the world
              </p>
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
